src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1640,66 +1494,150 @@ ExtractInsertPartitionValue(Query *query, Var *partitionColumn)
 static Job *
 RouterSelectJob(Query *originalQuery, RelationRestrictionContext *restrictionContext,
-				bool *returnQueryRoutable)
+				DeferredErrorMessage **planningError)
 {
 	Job *job = NULL;
 	Task *task = NULL;
-	bool queryRoutable = false;
 	StringInfo queryString = makeStringInfo();
 	uint64 shardId = INVALID_SHARD_ID;
 	List *placementList = NIL;
 	List *relationShardList = NIL;
 	bool replacePrunedQueryWithDummy = false;
+	bool requiresMasterEvaluation = false;
+	List *rangeTableList = NIL;
+	ListCell *rangeTableCell = NULL;
+	bool hasNullifiedModification = false;
 
 	/* router planner should create task even if it deosn't hit a shard at all */
 	replacePrunedQueryWithDummy = true;
 
-	queryRoutable = RouterSelectQuery(originalQuery, restrictionContext,
-									  &placementList, &shardId, &relationShardList,
-									  replacePrunedQueryWithDummy);
+	/* check if this query requires master evaluation */
+	requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
 
-
-	if (!queryRoutable)
+	RouterQuery(originalQuery, restrictionContext,
+				&placementList, &shardId, &relationShardList,
+				replacePrunedQueryWithDummy, planningError);
+	if (*planningError)
 	{
-		*returnQueryRoutable = false;
 		return NULL;
 	}
 
 	job = CreateJob(originalQuery);
 
+	/* extract range table entries */
+	ExtractRangeTableEntryWalker((Node *) originalQuery, &rangeTableList);
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		bool requiresUpdateOrDeletePermissions = false;
+
+		/* XXX: todo refactor into a function */
+		if ((ACL_UPDATE & rangeTableEntry->requiredPerms) ||
+			(ACL_DELETE & rangeTableEntry->requiredPerms))
+		{
+			requiresUpdateOrDeletePermissions = true;
+		}
+
+		if (rangeTableEntry->rtekind == RTE_SUBQUERY && requiresUpdateOrDeletePermissions)
+		{
+			hasNullifiedModification = true;
+			break;
+		}
+	}
+
+	/* If there are no shards for update, create an empty task list for modifications */
+	if (hasNullifiedModification)
+	{
+		job->taskList = NIL;
+		return job;
+	}
+
 	pg_get_query_def(originalQuery, queryString);
 
-	task = CreateTask(ROUTER_TASK);
+	if (originalQuery->commandType == CMD_SELECT)
+	{
+		task = CreateTask(ROUTER_TASK);
+	}
+	else
+	{
+		char replicationModel = REPLICATION_MODEL_INVALID;
+		bool distributedSelect = false;
+		char modificationPartitionMethod = 0;
+
+		/* XXX: todo refactor into a function */
+		foreach(rangeTableCell, rangeTableList)
+		{
+			RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+			Oid relationId = rangeTableEntry->relid;
+			DistTableCacheEntry *cacheEntry = NULL;
+			bool requiresUpdateOrDeletePermissions = false;
+
+			if (relationId == InvalidOid)
+			{
+				continue;
+			}
+
+			if ((ACL_UPDATE & rangeTableEntry->requiredPerms) ||
+				(ACL_DELETE & rangeTableEntry->requiredPerms))
+			{
+				requiresUpdateOrDeletePermissions = true;
```

Comment:
this loop would be a lot clearer if it was broken down in a few steps / separate loops,

e.g. thinking out lout:
```
modificationRTE = GetModificationRTE(rangeTableList);
modificationTableCacheEntry = DistributedTableCacheEntry(rangeTableEntry->relid);

modificationPartitionMethod = cacheEntry->partitionMethod;
if (modificationPartitionMethod == DISTRIBUTE_BY_NONE && SelectsFromDistributedTable(rangeTableList))
{
    ereport(ERROR, ... );
}

task = CreateTask(MODIFY_TASK);
task->replicationModel = cacheEntry->replicationModel;
```